Automate start of QEMU

svlad-90 opened the pull request:
domu: Introduce services for auto start of QEMU
In case of usage of the virtio functionality, we need a set of systemd
services, which would:

1. domu-create.service: start DomU guest domain, pause it, and send
formed QEMU command from Dom0 to DomD
2. domu-unpause.service: wait for QEMU to be started in DomD, and then
unpause the DomU guest domain
3. domu-restart-monitor: in case if DomU domain went into reboot, or was
destroyed - shut down first 2 services in order to cause the whole
start-up chain of the DomU domain ( including QEMU ) to be played from
the beginning

This patch introduces listed services, and applies them only in case if
'enable_virtio' distro feature was activated.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

backend-ready.bb: Fix typo in RDEPENDS directive
Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

backend-ready: introduce continuous observation
To synchronize start of the QEMU with the start of the guest domain we
need to have a backend-ready abstraction, which is capable of continuous
observation of the status, which is stored in the xen-store argument.

This patch changes the behavior of the backend-ready abstraction from
the 'one time ready notification' to the 'continuous observation'.

The major changes are:
- backend-ready service dies in case of "dead" notification from the
xen-store
- backend-ready service restarts itself in order to continue the
observation

This change should not impact existing users of this abstraction as
client side of this mechanism stays the same - based on the systemd
dependencies.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

backend-ready: Dom0->DomD communication
To sync QEMU startup with the startup of the guest domain we need to
 have 2 sides communication - DomD->Dom0 and Dom0->DomD. DomD->Dom0
communication already exists. Dom0->DomD - doesn't.

This patch introduces backend-ready abstraction specific for DomD
domain to cover the Dom0->DomD communication.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

virtio: Introduce qemu-restart-monitor.service
We need to recover from the cases, in which QEMU becomes unavailable,
e.g. due to its crash. In this case the guest domain should be
restarted, and the whole "start and pause guest domain -> start qemu ->
unpause guest domain" chain should be played from the beginning.

This patch introduces the qemu-restart-monitor systemd service, which
serves the above purpose.

It was not possible to re-use backend-ready abstraction for this task,
as this service should start independently from the other targets
during the boot. And achieving that for the systemd template instance
services from the Yocto context is quite complex.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

doma: Introduce services for auto start of QEMU
In case of usage of the virtio functionality, we need a set of systemd
services, which would:

1. doma-create.service: start DomA guest domain, pause it, and send
formed QEMU command from Dom0 to DomD
2. doma-unpause.service: wait for QEMU to be started in DomD, and then
unpause the DomA guest domain
3. doma-restart-monitor: in case if DomA domain went into reboot, or was
destroyed - shut down first 2 services in order to cause the whole
start-up chain of the DomA domain ( including QEMU ) to be played from
the beginning

This patch introduces listed services, and applies them only in case if
'enable_virtio' distro feature was activated.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

doma: Introduce services for auto start of QEMU
In case of usage of the virtio functionality, we need a set of systemd
services, which would:

1. doma-create.service: start DomA guest domain, pause it, and send
formed QEMU command from Dom0 to DomD
2. doma-unpause.service: wait for QEMU to be started in DomD, and then
unpause the DomA guest domain
3. doma-restart-monitor: in case if DomA domain went into reboot, or was
destroyed - shut down first 2 services in order to cause the whole
start-up chain of the DomA domain ( including QEMU ) to be played from
the beginning

This patch introduces listed services, and applies them only in case if
'enable_virtio' distro feature was activated.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

----

domu: Introduce services for auto start of QEMU
In case of usage of the virtio functionality, we need a set of systemd
services, which would:

1. domu-create.service: start DomU guest domain, pause it, and send
formed QEMU command from Dom0 to DomD
2. domu-unpause.service: wait for QEMU to be started in DomD, and then
unpause the DomU guest domain
3. domu-restart-monitor: in case if DomU domain went into reboot, or was
destroyed - shut down first 2 services in order to cause the whole
start-up chain of the DomU domain ( including QEMU ) to be played from
the beginning

This patch introduces listed services, and applies them only in case if
'enable_virtio' distro feature was activated.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>

svlad-90 commented:
The changes were verified in the following way:

|Build type|Comment|
|---|---|
|Non-virtio DomU|Boot is successful. Graphics works.|
|Non-virtio DomA|Boot is successful. Graphics works.|
|Virtio DomU|All expected systemd services are running. The system does not boot due to the lack of the proper Xen-configuration.|
|Virtio DomA|All expected systemd services are running. The system does not boot due to the lack of the proper Xen-configuration.|

lorc commented on meta-xt-control-domain/recipes-guest/doma/files/doma-restart-monitor-ExecStart.sh:
```diff
@@ -0,0 +1,59 @@
+DOMA_ID="";
+
+while [ "$DOMA_ID" == "" ]; do
+    DOMA_ID=$(xl list | awk '{ if ($1 == "DomA") print $2 }');
```
`xl domid`

svlad-90 replied:
Hi @lorc, I'm not sure that it will be better.

```
DOMD_ID="";

output=$(xl domid DomD)
if [ $? -eq 0 ]; then
    DOMD_ID=$output
fi
```

Also, how would it work with "set -e"?

svlad-90 replied:
Hi @lorc, I'll close this point as of now. In case of usage of the "xl domid" command, I'll need to additionally check for the error state (which happens when the domain is not available ) to void assigning the error content to the DOMD_ID variable.

The error looks like this:

```
root@generic-armv8-xt-dom0:~# xl domid non_existing_domain_id
Can't get domid of domain name 'non_existing_domain_id', maybe this domain does not exist.
```

It'll take more lines of code and won't be that elegant.

awk, in its turn, will return an empty string in case of a non-existing domain. That is more suitable in this specific case.

lorc commented on meta-xt-control-domain/recipes-guest/doma/files/doma-restart-monitor.service:
```diff
@@ -0,0 +1,15 @@
+
+ExecStart=/usr/lib/xen/bin/doma-restart-monitor-ExecStart.sh
+
+ExecStopPost=echo "DomA restart monitor stopped!"
```
Do you really need this?

svlad-90 replied:
Hi @lorc, done.

lorc commented on meta-xt-control-domain/recipes-guest/doma/files/doma-restart-monitor.service:
```diff
@@ -0,0 +1,15 @@
+ExecStopPost=echo "DomA restart monitor stopped!"
+
+Restart=on-failure
+RemainAfterExit=true
```
> RemainAfterExit=
>Takes a boolean value that specifies whether the service shall be considered active even when all its processes exited.

Is this a correct option for this service?

svlad-90 replied:
Hi @lorc, I think it is not needed here, as the script of this service remains active until doma restart.

Accepted.

svlad-90 replied:
Hi @lorc,

Done for doma-restart-monitor.service and domu-restart-monitor.service

lorc commented on meta-xt-control-domain/recipes-guest/doma/files/doma-unpause.service:
```diff
@@ -0,0 +1,17 @@
+ExecStart=/bin/bash -c 'echo "[doma-unpause]: Unpausing DomA" && /usr/sbin/xl -v unpause DomA && echo "[doma-unpause]: Unpaused DomA"'
+
+Restart=always
+RemainAfterExit=true
```
> RemainAfterExit=
>Takes a boolean value that specifies whether the service shall be considered active even when all its processes exited.

Is this a correct option for this service?

svlad-90 replied:
Hi @lorc,

Done for doma-unpause.service and domu-unpause.service

lorc commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create.service:
```diff
@@ -0,0 +1,39 @@
+ExecStopPost=/usr/lib/xen/bin/doma-create-ExecStopPost.sh
+
+Restart=always
+RemainAfterExit=true
```
> RemainAfterExit=
>Takes a boolean value that specifies whether the service shall be considered active even when all its processes exited.

Is this a correct option for this service? I have no idea how it should work with "Restart=always"

svlad-90 replied:
Hi @lorc, Here it is totally valid. It will remain valid even after the squash of doma-create and doma-unpause services.

The script of this service does something, and then, at some point, the script finishes and exits. Still, the service should remain active until DomA is running, as we are interested in the execution of the ExecStopPost in case if user calls "systemctl stop doma-create".

svlad-90 replied:
Hi @lorc, regarding this point I find nothing to do, and I'm pretty convinced in a made conclusion. Thus, I'll close this conversation.

Please, reopen it if there are still open questions.

lorc commented on meta-xt-control-domain/recipes-guest/domu/files/domu-create-ExecStart.sh:
```diff
@@ -0,0 +1,44 @@
+-device virtio-net-pci,disable-legacy=on,iommu_platform=on,romfile=\"\",id=nic0,netdev=net0,mac=08:00:27:ff:cb:cf \
+-netdev type=tap,id=net0,ifname=vif-emu,br=xenbr0,script=no,downscript=no \
+-device virtio-blk-pci,scsi=off,disable-legacy=on,iommu_platform=on,drive=image \
+-drive if=none,id=image,format=raw,file=/dev/mmcblk0p3 \
```
In many cases we boot DomU via NFS. How this is handled in your case?

svlad-90 replied:
Hi @lorc, accepted and done.

I've changed the implementation to support DomU boot via the NFS. It works.

The change required additional commit to the meta-xt-prod-devel-rcar:
https://github.com/xen-troops/meta-xt-prod-devel-rcar/pull/101/commits/ddad8aab2a9942d5932dfee7a32bb8daec756fb8

Done.

lorc reviewed (changes requested):
Is there any particular reason why do you need -create and -unpase services?

As I see this: doma.service starts a bash script that
1. Creates paused domain using `xl create -p`
2.  Signals QEMU on other end
3. Calls wait-backend script to wait while QEMU is started
4. Unpauses domain. 
5. PROFIT

lorc commented:
Also, could you please write in the commit message that this monstrosity is temporary? 

Also, this is meta-xt-common. Why do you have hardcoded settings for DomU and DomA here? I am talking about IP addresses and block device names (/dev/mmcblk0p3)

svlad-90 commented:
> Is there any particular reason why do you need -create and -unpase services?
> 
> As I see this: doma.service starts a bash script that
> 
> 1. Creates paused domain using `xl create -p`
> 2. Signals QEMU on other end
> 3. Calls wait-backend script to wait while QEMU is started
> 4. Unpauses domain.
> 5. PROFIT

Hi @lorc, the backend-ready abstraction is the reason to do it in such a way.

I'll describe it in terms of doma-create and doma-unpause systemd services.

doma-unpause.service requires backend-ready@domd-qemu-monitor.service. That dependency becomes available only after doma-create writes the required content to the xen-store.

If I'll continue to operate in terms of backend-ready abstraction I won't be able to eliminate doma-unpause service, as I'll need to create the following "Requies" section in doma-create:

```
Requires=backend-ready@block.service backend-ready@domd-qemu-monitor.service ...
```

With such "Requires" section doma-create.service ( or doma.service, if I rename it ) will never be started, as backend-ready@domd-qemu-monitor.service dependency will never be fulfilled.

So doing it in another way means switching to another approach.

----

It is possible to switch to the sequence, proposed by you.

But step 3 is wrong, as the current implementation of the wait-for-backend script does not end on availability notification. It continues observation until the backend becomes unavailable, and then fails by intention. I'll rename wait-for-backend to observe-backend to make it more obvious. Anyway, using the script directly will block the control flow. Instead of re-using the script, I'll need to copy-paste part of it into the virtio version of the doma.service. 

For the same reason, step 5 "PROFIT" is not sufficient. Someone should still observe the QEMU availability and stop the service in case of a QEMU crash. It means that I will need to replace the usage of the backend-ready@domd-qemu-monitor.service with some amount of copy-paste code from the wait-for-backend script for this part.

----

Here we have a tradeoff:
1. Have additional systemd service, but operate at a higher level of abstraction. Implementation is more transparent, but at a price of systemd services structure difference between the virtio and non-virtio builds
2. Have fewer services, but with more copy-pasted bash script code. It will lead to dependency handling partially on systemd level and partially on the bash scripting level. I would like to have such handling on a single level wherever possible.

I prefer option #1. We already have differences in the structure of the systemd services between virtio and non-virtio builds. E.g. displbe.service, sndbe.service, displaymanager.service. So, having one more difference is not something new.

Considering the above, and the weeks, that were spent debugging and stabilizing this solution during execution of CTS tests, I would prefer to leave it as it is - doma-create.service + doma-unpause.service.

To compensate existence of such differences, let me document major differences between virtio and non-virtio builds in the README.md of meta-xt-common and meta-xt-prod-devel-rcar.

As this monstrosity is temporary, keep in mind that someday it will go away including the difference we are discussing now.

----

Also, regarding renaming doma-create.service into doma.service. I was thinking of this but came to the conclusion that if we are leaving the pair of services - doma-create.service + doma-unpause.service - it would be better to have the doma-create.service name instead of doma.service. It'll emphasize, that there is a difference, leaving less room for confusion.

svlad-90 commented:
> > Is there any particular reason why do you need -create and -unpase services?
> > As I see this: doma.service starts a bash script that
> > 
> > 1. Creates paused domain using `xl create -p`
> > 2. Signals QEMU on other end
> > 3. Calls wait-backend script to wait while QEMU is started
> > 4. Unpauses domain.
> > 5. PROFIT
> 
> Hi @lorc, the backend-ready abstraction is the reason to do it in such a way.
> 
> I'll describe it in terms of doma-create and doma-unpause systemd services.
> 
> doma-unpause.service requires [backend-ready@domd-qemu-monitor.service](mailto:backend-ready@domd-qemu-monitor.service). That dependency becomes available only after doma-create writes the required content to the xen-store.
> 
> If I'll continue to operate in terms of backend-ready abstraction I won't be able to eliminate doma-unpause service, as I'll need to create the following "Requies" section in doma-create:
> 
> ```
> Requires=backend-ready@block.service backend-ready@domd-qemu-monitor.service ...
> ```
> 
> With such "Requires" section doma-create.service ( or doma.service, if I rename it ) will never be started, as [backend-ready@domd-qemu-monitor.service](mailto:backend-ready@domd-qemu-monitor.service) dependency will never be fulfilled.
> 
> So doing it in another way means switching to another approach.
> 
> It is possible to switch to the sequence, proposed by you.
> 
> But step 3 is wrong, as the current implementation of the wait-for-backend script does not end on availability notification. It continues observation until the backend becomes unavailable, and then fails by intention. I'll rename wait-for-backend to observe-backend to make it more obvious. Anyway, using the script directly will block the control flow. Instead of re-using the script, I'll need to copy-paste part of it into the virtio version of the doma.service.
> 
> For the same reason, step 5 "PROFIT" is not sufficient. Someone should still observe the QEMU availability and stop the service in case of a QEMU crash. It means that I will need to replace the usage of the [backend-ready@domd-qemu-monitor.service](mailto:backend-ready@domd-qemu-monitor.service) with some amount of copy-paste code from the wait-for-backend script for this part.
> 
> Here we have a tradeoff:
> 
> 1. Have additional systemd service, but operate at a higher level of abstraction. Implementation is more transparent, but at a price of systemd services structure difference between the virtio and non-virtio builds
> 2. Have fewer services, but with more copy-pasted bash script code. It will lead to dependency handling partially on systemd level and partially on the bash scripting level. I would like to have such handling on a single level wherever possible.
> 
> I prefer option 1. We already have differences in the structure of the systemd services between virtio and non-virtio builds. E.g. displbe.service, sndbe.service, displaymanager.service. So, having one more difference is not something new.
> 
> Considering the above, and the weeks, that were spent debugging and stabilizing this solution during execution of CTS tests, I would prefer to leave it as it is - doma-create.service + doma-unpause.service.
> 
> To compensate existence of such differences, let me document major differences between virtio and non-virtio builds in the README.md of meta-xt-common and meta-xt-prod-devel-rcar.
> 
> As this monstrosity is temporary, keep in mind that someday it will go away including the difference we are discussing now.
> 
> Also, regarding renaming doma-create.service into doma.service. I was thinking of this but came to the conclusion that if we are leaving the pair of services - doma-create.service + doma-unpause.service - it would be better to have the doma-create.service name instead of doma.service. It'll emphasize, that there is a difference, leaving less room for confusion.

Hi @lorc, I've added an additional commit with an update of the documentation to the meta-xt-prod-devel-rcar repository:
https://github.com/xen-troops/meta-xt-prod-devel-rcar/pull/101/commits/f39d01dc29cd643bccce9782bbbe11a0af49bb69

Also, I've renamed wait-for-backend into the observe-backend in the corresponding commits of this PR.

lorc commented:
> I prefer option 1. We already have differences in the structure of the systemd services between virtio and non-virtio builds. E.g. displbe.service, sndbe.service, displaymanager.service. So, having one more difference is not something new.

Well, backend-ready@.service was designed with classical Xen PV backends in mind. No need to try to use this tool for virtio/qemu, as this is not intended use-case. 

> Considering the above, and the weeks, that were spent debugging and stabilizing this solution during execution of CTS tests, I would prefer to leave it as it is - doma-create.service + doma-unpause.service.

I would prefer to have something manageable. Your current approach involves to many moving parts. 

> But step 3 is wrong, as the current implementation of the wait-for-backend script does not end on availability notification. It continues observation until the backend becomes unavailable, and then fails by intention.

You can just add command line parameter to make it exit after back-end becomes ready. You already reworking this script, so I see no problem here.

> Considering the above, and the weeks, that were spent debugging and stabilizing this solution during execution of CTS tests, I would prefer to leave it as it is - doma-create.service + doma-unpause.service.

This is a nature of RnD: at first you are finding some working solution and then you are developing proper implementation. As 100% proper fix (with xen toolstack) is not expected in the nearest future, I would like to see a properly engineered workaround. 

Also, looks like you ignored all my other comments.

svlad-90 commented:
Hi @lorc,

> > Well, [backend-ready@.service](mailto:backend-ready@.service) was designed with classical Xen PV backends in mind. No need to try to use this tool for virtio/qemu, as this is not intended use-case.

I do not agree with you regarding this point. The time may come when this or that abstraction can be re-used for the additional use case. I've adapted backend-ready to serve both - old and new use cases. I think that the reuse of code is a preferable way.

> > I would prefer to have something manageable. Your current approach involves too many moving parts.

I would prefer to finish the task without the need to reimplement it once again. You ask for a huge change that might require additional days or weeks of debugging. It was really hard to test all the different cases. I fear that I'll break something and will not even notice. To catch some of the cases I needed to run CTS on Android and wait for hours to reproduce e.g. specific crash of QEMU, and test that I'm able to recover.

I do not reject the requested changes. Just try to force me to accept them, as it would mean more effort than you probably think. Or I'll skip testing some part of the rare use-cases, as a compromise. ))

> > You can just add command line parameter to make it exit after back-end becomes ready. You already reworking this script, so I see no problem here.

That is possible. I'll think about whether it will help a lot or not. In this case, I'm more concerned about the need to call the wait-for-backend script directly from another sh script, used by a systemd service. Looks like a break of encapsulation.

> > This is a nature of RnD: at first you are finding some working solution and then you are developing proper implementation. As 100% proper fix (with xen toolstack) is not expected in the nearest future, I would like to see a properly engineered workaround.

As I understand, your main concern is that we should keep a single doma(u).service instead of the doma(u)-create.service + doma(u)-unpause.service pair. I think I have no other choice than to accept it. But keep in mind, that it will take some time.

> > Also, looks like you ignored all my other comments.

I've pushed part of the changes, to try to convince you regarding this major topic. So other topics were out of scope until now. I'll answer to them right now. Sorry for the delay.

svlad-90 commented on meta-xt-control-domain/recipes-extended/backend-ready/files/wait-for-backend:
```diff
@@ -10,15 +10,27 @@ BACKEND=$1
 while true; do
-    xenstore-watch -n2 $XS_PATH&
-    WATCH_PID=$!
+    if [ x$CURRENT_STATUS = x"unknown" ]; then
```
Hi @lorc, your approach was checked. Accepted.

svlad-90 replied:
Hi @lorc, Done.

svlad-90 commented on meta-xt-control-domain/recipes-extended/qemu-restart-monitor/files/qemu-restart-monitor.service:
```diff
@@ -0,0 +1,15 @@
+
+ExecStart=/usr/lib/xen/bin/qemu-restart-monitor-ExecStart.sh
+
+ExecStop=echo "QEMU restart monitor stopped!"
```
Hi @lorc, Yes, I really need this. There is a counterpart of this line here:

https://github.com/xen-troops/meta-xt-common/pull/62/files#diff-048135201e409d8fb8074ace3cab8c019468a070c364caf61eda3b1c552c1344R3

And the same with other traces in this PR.

It is done to increase readability. I've tried to proceed without these additional logs. It took more time for me to go through the scenario in journalctl without them.

Example from the target system:

```
vladyslav_goncharuk@EPUAODEW00C5:~/Documents/Junk/debug_new_kernel$ cat 10_May_2023_qemu_restart.txt | grep qemu-restart-monitor
Apr 28 17:42:26 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[192]: QEMU restart monitor started!
Apr 28 17:42:29 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[192]: Parsed DOMD_ID is '1'
Apr 28 17:42:29 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[252]: xenstore-read: couldn't read path /local/domain/1/drivers/domd-qemu-monitor/status
Apr 28 17:42:29 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[192]: QEMU status is ''.
Apr 28 17:43:02 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[192]: QEMU status is 'ready'.
Apr 28 17:54:01 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[192]: QEMU status is 'dead'.
Apr 28 17:54:01 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[192]: Failing to notify QEMU unavailability. Will be restarted soon ...
Apr 28 17:54:01 generic-armv8-xt-dom0 systemd[1]: qemu-restart-monitor.service: Main process exited, code=exited, status=1/FAILURE
Apr 28 17:54:01 generic-armv8-xt-dom0 systemd[1]: qemu-restart-monitor.service: Failed with result 'exit-code'.
Apr 28 17:54:07 generic-armv8-xt-dom0 systemd[1]: qemu-restart-monitor.service: Scheduled restart job, restart counter is at 1.
Apr 28 17:54:09 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[454]: QEMU restart monitor started!
Apr 28 17:54:09 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[454]: Parsed DOMD_ID is '1'
Apr 28 17:54:09 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[454]: QEMU status is 'dead'.
Apr 28 17:54:23 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[454]: QEMU status is 'ready'.
```

svlad-90 commented on meta-xt-control-domain/recipes-extended/qemu-restart-monitor/files/qemu-restart-monitor-ExecStart.sh:
```diff
@@ -0,0 +1,40 @@
+# Fetch DomD Xen domain identifier
+DOMD_ID="";
+
+while [ "$DOMD_ID" == "" ]; do
```
Hi @lorc, accepted. I will try that out.

svlad-90 replied:
Hi @lorc, I've tried your solution.

I've added the following line to the qemu-restart-monitor.service's Unit section:

```
Requires=domd.service
```

and modified qemu-restart-monitor-ExecStart.sh in the following way:

```
# Fetch DomD Xen domain identifier
DOMD_ID=$(xl domid DomD);
echo "Parsed DOMD_ID is '${DOMD_ID}'";
```

I'm getting such an error:

````
Apr 28 17:42:26 generic-armv8-xt-dom0 qemu-restart-monitor-ExecStart.sh[139]: Can't get domid of domain name 'DomD', maybe this domain does not exist.
```

domd.service is a oneshot service. Thus, probably, it is considered as "started" even before "xl create" is finished. In this case, qemu-restart-monitor with some probability will be started before the DomD id is added to the "xl list".

I can adjust domd.service to become a simple service, but that will be nonsense. I don't want to have my implementation relying on type of some other service. I want it to be robust enough to survive as many unpredictable situations as possible. I will my implementation as is, as it is much more robust than what was proposed.

svlad-90 commented on meta-xt-driver-domain/recipes-extended/qemu-command-handler/qemu-command-handler.bb:
```diff
@@ -0,0 +1,37 @@
+SUMMARY = "Service for the block device up notification"
```
Hi @lorc, copy-paste issue. Accepted.

svlad-90 replied:
Hi @lorc, done.

svlad-90 commented on meta-xt-driver-domain/recipes-extended/qemu-command-handler/qemu-command-handler.bb:
```diff
@@ -0,0 +1,37 @@
+"
+
+RRECOMMENDS:${PN} += " \
+    virtual/xenstored \
```
Hi @lorc, I already do not remember, why that was added. I'll try without it. If it will work - I'll remove it. If not - I'll come back with additional status.

svlad-90 replied:
Hi @lorc, done. Only xen-tools-xenstore has left.

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create-ExecStart.sh:
```diff
@@ -0,0 +1,44 @@
+
+echo "Creating DomA" && /usr/sbin/xl -v create /etc/xen/doma.cfg && \
+echo "Created DomA" && \
+echo "Pausing DomA" && /usr/sbin/xl -v pause DomA && \
```
Hi @lorc, it was not working for a while, so I've decided to decouple each operation and explicitly state, that it has finished successfully. As now it works, there is no need to have that separation anymore. Accepted.

svlad-90 replied:
Hi @lorc, done for both DomA and DomU.

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create-ExecStart.sh:
```diff
@@ -0,0 +1,44 @@
+
+echo "Sleeping 5 seconds" && sleep 5;
+
+echo "Creating DomA" && /usr/sbin/xl -v create /etc/xen/doma.cfg && \
```
Hi @lorc,

I assume you mean "set -e", right?

Anyway, in this specific script ( and the same for domu ) I can try it out.

Accepted. I will also review the other files.

svlad-90 replied:
Hi @lorc,

Done for doma-create-ExecStart.sh and domu-create-ExecStart.sh

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create-ExecStart.sh:
```diff
@@ -0,0 +1,44 @@
+echo "Created DomA" && \
+echo "Pausing DomA" && /usr/sbin/xl -v pause DomA && \
+echo "Paused DomA" && \
+DOMD_ID=$(xl list | awk '{ if ($1 == "DomD") print $2 }') && \
```
Hi @lorc, accepted.

svlad-90 replied:
Hi @lorc,

Done for both doma-create-ExecStart.sh and domu-create-ExecStart.sh

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create-ExecStopPost.sh:
```diff
@@ -0,0 +1,40 @@
+# of DomD will cause qemu-command-handler service in DomD to be
+# stopped. That will unblock shutdown of the DomA guest domain.
+
+DOMD_ID=$(xl list | awk '{ if ($1 == "DomD") print $2 }') && \
```
Hi @lorc, accepted.

svlad-90 replied:
Hi @lorc, done for both DomA and DomU.

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create-ExecStopPost.sh:
```diff
@@ -0,0 +1,40 @@
+HANGING_DOMAIN_ID_MEMORIZED="";
+
+while true; do
+    HANGING_DOMAIN_ID=$(xl list | awk '{ if ($1 == "(null)") print $2 }');
```
Hi @lorc, We have DOMD_ID here. But we are destroying DomA.

After DomA is destroyed, it often happens that it is still stuck in "xl list" with its name equal to "(null)". Probably domain is releasing its resources. I don't know for sure. It can happen for dozens of seconds.

Here I'm waiting for the domain to totally disappear from the "xl list" because otherwise, it might happen that (null) domain still consumes memory, and the creation of a new instance of DomA will fail in such case. It actually has happened.

Also, it could happen, that when I reach this script the domain was already destroyed ( e.g. internal reboot of the domain ). So getting its ID here was not an option.

That's why it is so tricky.

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-restart-monitor-ExecStart.sh:
```diff
@@ -0,0 +1,59 @@
+XS_PATH="/local/domain/$DOMA_ID"
+
+# Fetch first availability of the parameter
+until xenstore-read $XS_PATH; do
```
Hi @lorc, it will not work, as here I'm waiting for the availability of the root XS_PATH itself. It is not a leaf node of the xen-store. No one assigns any values to it, so xenstore-watch will stuck forever.

```
# Trying to work with non-running domain 3
# xenstore-watch -n 1 will return without errors independently from whether the path already exists or not. So it is not representative
# xenstore-watch -n 2 will stuck forever in case this service was started after the domain was already running. So also will not work to check when the domain has become available
root@generic-armv8-xt-dom0:~# xenstore-watch -n 1 /local/domain/3
/local/domain/3
# read operation will fail if there is no path existing. Can be used for the check. 
root@generic-armv8-xt-dom0:~# xenstore-read /local/domain/3                                                                                                                        
xenstore-read: couldn't read path /local/domain/3
```

svlad-90 commented:
> > Also, could you please write in the commit message that this monstrosity is temporary?

> > Also, this is meta-xt-common. Why do you have hardcoded settings for DomU and DomA here? I am talking about IP addresses and block device names (/dev/mmcblk0p3)

Hi @lorc, regarding commit messages - accepted.
Regarding the hardcoded device name - accepted. Will change it to be injected on the Yocto level.
Regarding IP address - usage of NFS has nothing to do with sharing of the device over the QEMU. I'll make sure that during usage of NFS QEMU doesn't share the block device with the DomU.

svlad-90 commented on meta-xt-control-domain/recipes-extended/qemu-restart-monitor/files/qemu-restart-monitor-ExecStart.sh:
```diff
@@ -0,0 +1,40 @@
+DOMD_ID="";
+
+while [ "$DOMD_ID" == "" ]; do
+    DOMD_ID=$(xl list | awk '{ if ($1 == "DomD") print $2 }');
```
Hi @lorc, I'll close this point as of now. In case of usage of the "xl domid" command, I'll need to additionally check for the error state (which happens when the domain is not available ) to void assigning the error content to the DOMD_ID variable.

The error looks like this:

```
root@generic-armv8-xt-dom0:~# xl domid non_existing_domain_id
Can't get domid of domain name 'non_existing_domain_id', maybe this domain does not exist.
```

It'll take more lines of code and won't be that elegant.

awk, in its turn, will return empty string in case of non-existing domain.

svlad-90 commented on meta-xt-control-domain/recipes-guest/doma/files/doma-create-ExecStart.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+echo "Sleeping 5 seconds" && sleep 5;
```
Hi @lorc, I've tested all domu and doma scenarios and restart was fine even without this sleep. Thus, removed.

Done.

svlad-90 commented:
Hi @lorc, status update.

I've fixed or replied to all the issues except squashing doma(u)-create and doma(u)-unpause. I hope to finish that during the weekend and have all PRs in the finalized state on Monday.

Also, at the current point, DomA & DomU with and without virtio are booting and starting successfully. Including booting DomU via NFS.

svlad-90 commented:
Hi @lorc, I've squashed doma(u)-create.service and doma(u)-unpause.service into the doma(u)-virtio service, which is installed into the system as doma(u).service in case of virtio build.

The PR is ready to be reviewed again. Please, have a look.

svlad-90 commented:
Another PR will be created instead of this one, as this is already overloaded with comments.
